fortitude/src/rules/style/functions.rs

```diff
@@ -0,0 +1,103 @@
+use crate::settings::Settings;
+use crate::{AstRule, FromAstNode};
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_source_file::SourceFile;
+use tree_sitter::Node;
+
+/// ## What it does
+///
+/// Checks that every function explicitly declares which variable contains the result.
+///
+/// ## Why is this bad?
+/// Specifying the `result()` clause allows for the variable inside the function containing the
+/// result to be named differently from the function.
+///
+/// Allowing the variable to be named different from the function allows for better naming of both
+/// the function and the internal variables, and also it can help when creating functions that might
+/// need to be duplicated to have versions supporting various types. Since declaring the result in
+/// the `result()` clause allows for easier duplication of the function because then less lines must
+/// be changed.
+///
+/// ## Example
+/// ```f90
+/// interface distance
+///     module procedure distance_real32, distance_real64
+/// end interface distance
+///
+/// function distance_real32(a, b)
+///     real(kind=real32), intent(in) :: a, b
+///     real(kind=real32)             :: distance_real32
+///
+///     distance_real32 = abs(a-b)
+/// end function distance_real
+///
+/// function distance_real64(a, b)
+///     real(kind=real64), intent(in) :: a, b
+///     real(kind=real64)             :: distance_real64
+///
+///     distance_real64 = abs(a-b)
+/// end function distance_real
+/// ```
+///
+/// Use instead:
+/// ```f90
+/// interface distance
+///     module procedure distance_real32, distance_real64
+/// end interface distance
+///
+/// function distance_real32(a, b) result(distance)
+///     integer, parameter               :: func_type = real32
+///
+///     real(kind=func_type), intent(in) :: a, b
+///     real(kind=func_type)             :: distance
+///
+///     distance = abs(a-b)
+/// end function distance_real
+///
+/// function distance_real64(a, b) result(distance)
+///     integer, parameter               :: func_type = real64
+///
+///     real(kind=func_type), intent(in) :: a, b
+///     real(kind=func_type)             :: distance
+///
+///     distance = abs(a-b)
+/// end function distance_real
+/// ```
+///
+/// ## References
+/// - Clerman, N. Spector, W., 2012, _Modern Fortran: Style and Usage_, Cambridge
+///   University Press, Rule 129, p. 178.
+#[derive(ViolationMetadata)]
+pub(crate) struct FunctionMissingResult;
+
+impl Violation for FunctionMissingResult {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Function missing result() specifier".to_string()
```

Comment:
Would it be possible to include the function name in the help text? `rules/style/end_statements.rs` has a similar example to this.

Reply:
The benefit of putting the function name in the title is that it gives a little more info when users are using `--output-format=concise` (or similar), but, as you've pointed out with `unnamed-end-statement`, we haven't always been very consistent about this (sorry, I didn't spot that it was just in the fix title!).

I'd be happy to merge with things as they are now. We're planning to do a big overhaul of how AST rules are handled soon in order to make it easier to keep track of scoping rules and other useful information, so a lot of rules will likely get minor updates like this as part of that process.